File: src/main/typescript/wcardinal/controller/internal/properties.ts
/*
 * Copyright (C) 2019 Toshiba Corporation
 * SPDX-License-Identifier: Apache-2.0
 */

import { Property } from "./property";
import { EnumSet } from "./settings";

export class Properties {
	private _properties: EnumSet<Property>;

	constructor( properties: EnumSet<Property> ) {
		this._properties = properties;
	}

	isReadOnly_(): boolean {
		return this.contains_( Property.READ_ONLY );
	}

	isNonNull_(): boolean {
		return this.contains_( Property.NON_NULL );
	}

	isSoft_(): boolean {
		return this.contains_( Property.SOFT );
	}

	isLocal_(): boolean {
		return this.contains_( Property.LOCAL );
	}

	isHistorical_(): boolean {
		return this.contains_( Property.HISTORICAL );
	}

	isAjax_(): boolean {
		return this.contains_( Property.AJAX );
	}

	add_( properties: EnumSet<Property> ): this {
		this._properties |= properties;
		return this;
	}

	remove_( properties: EnumSet<Property> ): this {
		this._properties &= ~properties;
		return this;
	}

	retain_( properties: EnumSet<Property> ): this {
		this._properties &= properties;
		return this;
	}

	toggle_( isOn: boolean, property: Property ): this {
		return ( isOn ? this.add_( property ) : this.remove_( property ) );
	}

	contains_( property: Property ) {
		return Properties.contains_( this._properties, property );
	}

	clone_() {
		return new Properties( this._properties );
	}

	static nonOf_(): Properties {
		return new Properties( Property.NONE );
	}

	static of_( a: Property ): Properties;
	static of_( a: Property, b: Property ): Properties;
	static of_( a: Property, b: Property, c: Property ): Properties;
	static of_( a: Property ) {
		const args = arguments;
		let properties: EnumSet<Property> = a;
		for( let i = 1, imax = args.length; i < imax; ++i ) {
			properties |= args[ i ];
		}
		return new Properties( properties );
	}

	static contains_( properties: EnumSet<Property>, property: Property ) {
		return (properties & property) !== 0;
	}
}

File: src/main/java/org/wcardinal/controller/internal/Properties.java
/*
 * Copyright (C) 2019 Toshiba Corporation
 * SPDX-License-Identifier: Apache-2.0
 */

package org.wcardinal.controller.internal;

import java.lang.annotation.Annotation;
import java.lang.reflect.AccessibleObject;
import java.util.EnumSet;

import org.wcardinal.controller.annotation.ReadOnly;
import org.wcardinal.controller.data.annotation.Historical;
import org.wcardinal.controller.data.annotation.NonNull;
import org.wcardinal.controller.data.annotation.Soft;
import org.wcardinal.controller.data.annotation.Uninitialized;

public class Properties {
	static boolean is( final Class<?> type, final Class<? extends Annotation> annotationClass ) {
		return type.isAnnotationPresent(annotationClass);
	}

	static boolean is( final AccessibleObject object, final Class<? extends Annotation> annotationClass ) {
		return object.isAnnotationPresent(annotationClass);
	}

	static void addIf( final EnumSet<Property> properties, final boolean check, final Property property ) {
		if( check ) properties.add( property );
	}

	static public EnumSet<Property> create(
			final boolean isShared,
			final boolean isFactory,
			final boolean isReadOnly,
			final boolean isHistorical,
			final boolean isNonNull,
			final boolean isUninitialized,
			final boolean isSoft
	) {
		final EnumSet<Property> result = empty();
		addIf( result, isShared, Property.SHARED );
		addIf( result, isFactory, Property.FACTORY );
		addIf( result, isReadOnly, Property.READ_ONLY );
		addIf( result, isHistorical, Property.HISTORICAL );
		addIf( result, isNonNull, Property.NON_NULL );
		addIf( result, isUninitialized, Property.UNINITIALIZED );
		addIf( result, isSoft, Property.SOFT );
		return result;
	}

	static public EnumSet<Property> create( final AccessibleObject object ) {
		final EnumSet<Property> result = empty();
		addIf( result, is( object, ReadOnly.class ), Property.READ_ONLY );
		addIf( result, is( object, Historical.class ), Property.HISTORICAL );
		addIf( result, is( object, NonNull.class ), Property.NON_NULL );
		addIf( result, is( object, Uninitialized.class ), Property.UNINITIALIZED );
		addIf( result, is( object, Soft.class ), Property.SOFT );
		return result;
	}

	static public EnumSet<Property> create( final Class<?> type ) {
		final EnumSet<Property> result = empty();
		addIf( result, is( type, ReadOnly.class ), Property.READ_ONLY );
		addIf( result, is( type, Historical.class ), Property.HISTORICAL );
		addIf( result, is( type, NonNull.class ), Property.NON_NULL );
		addIf( result, is( type, Uninitialized.class ), Property.UNINITIALIZED );
		addIf( result, is( type, Soft.class ), Property.SOFT );
		return result;
	}

	static public EnumSet<Property> create( final EnumSet<Property> a, final EnumSet<Property> b ) {
		final EnumSet<Property> result = EnumSet.copyOf( a );
		result.addAll( b );
		return result;
	}

	static public EnumSet<Property> create( final EnumSet<Property> a, final Property b ) {
		final EnumSet<Property> result = EnumSet.copyOf( a );
		result.add( b );
		return result;
	}

	static public int toInt(EnumSet<Property> properties) {
		int result = 0;
		for( final Property property: properties ) {
			result |= (1 << property.ordinal());
		}
		return result;
	}

	static public EnumSet<Property> empty() {
		return EnumSet.noneOf(Property.class);
	}
}
